adservertargeting/reqdataprocessor.go

```diff
@@ -0,0 +1,104 @@
+package adservertargeting
+
+import (
+	"encoding/json"
+	"github.com/buger/jsonparser"
+	"github.com/pkg/errors"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"net/url"
+	"strings"
+)
+
+func getAdServerTargeting(reqWrapper *openrtb_ext.RequestWrapper) ([]openrtb_ext.AdServerTargeting, error) {
+	reqExt, err := reqWrapper.GetRequestExt()
+	if err != nil {
+		return nil, err
+	}
+	reqExtPrebid := reqExt.GetPrebid()
+	if reqExtPrebid == nil {
+		return nil, err
+	}
+	return reqExtPrebid.AdServerTargeting, nil
+}
+
+func getValueFromBidRequest(dataHolder *reqImpCache, path string, queryParams url.Values) (RequestTargetingData, error) {
+	//use the path specified in 'value' to look for data in the ortb bidrequest.
+	res := RequestTargetingData{}
+
+	//check if key points to query param from ext.prebid.amp.data
+	ampDataValue, err := getValueFromQueryParam(path, queryParams)
+	if ampDataValue != nil || err != nil {
+		res.SingleVal = ampDataValue
+		return res, err
+	}
+
+	// check if key points to imp data
+	impData, err := getValueFromImp(path, dataHolder)
+	if len(impData) > 0 || err != nil {
+		res.TargetingValueByImpId = impData
+		return res, err
+	}
+
+	// get data by key from request
+	requestValue, err := getDataFromRequest(path, dataHolder)
+	if requestValue != nil || err != nil {
+		res.SingleVal = requestValue
+		return res, err
+	}
+
+	return res, nil
+}
+
+func getValueFromQueryParam(path string, queryParams url.Values) (json.RawMessage, error) {
+	ampDataSplit := strings.Split(path, "ext.prebid.amp.data.")
```

Comment:
Using split in this manner is a bit confusing to read. Consider extracting this to a method like:

```
func verifyPrefixAndTrim(path, prefix string) (string, bool) {
...
}
```

then this call site looks like

```
pathTrimmed, hasPrefix := verifyPrefixAndTrim(path, "ext.prebid.amp.data.")
if !hasPrefix {
  return nil, nil
}

if val := queryParams.Get(pathTrimmed); val != "" {
  return json.RawMessage(val), nil
}

return nil, errors.Errorf("value not found for path: %s", path)
```

Do you think that's easier to follow with the details of the split length hidden in the helper func?

Reply:
Yes! Definitely looks better.